Exercises/NoModules/Chapter 13/Soln13_09/Truckload.cpp

```diff
@@ -0,0 +1,151 @@
+#include "Truckload.h"
+
+#include <iostream>
+
+// Definition of the nested class member
+// Since this member is private, 
+// its definition can be moved to the source file.
+class Truckload::Package
+{
+public:
+  SharedBox m_box;      // Pointer to the Box object contained in this Package
+  Package* m_next;      // Pointer to the next Package in the list
+
+  Package(SharedBox box) : m_box{ box }, m_next{ nullptr } {} // Constructor
+  ~Package() { delete m_next; }                           // Destructor
+};
+
+// Constructor - one Box (moved to source file to gain access to definition of Package)
+Truckload::Truckload(SharedBox box)          
+{
+  m_head = m_tail = new Package{ box };
+}
+
+// Constructor - vector of Boxes
+Truckload::Truckload(const std::vector<SharedBox>& boxes)
+{
+  for (const auto& box : boxes)
+  {
+    addBox(box);
+  }
+}
+
+// Copy constructor
+Truckload::Truckload(const Truckload& src)
+{
+  for (Package* package{ src.m_head }; package; package = package->m_next)
+  {
+    addBox(package->m_box);
+  }
+}
+
+Truckload& Truckload::operator=(const Truckload& other)
+{
+  if (&other != this)   // Do not forget: avoid issues with self-assignment!
+  {
+    delete m_head;              // Delete all current packages
+    m_head = m_tail = nullptr;  // Reset both pointers
+
+    // Same as copy constructor 
+    // (see Chapter 17 for the copy-and-swap idiom that allows you to avoid 
```

Comment:
It seems you added something to this chapter for copy-swap.
At least I see an Ex13_12B (which is not yet in my draft though) that explains copy-swap for Message.
If that's correct, then you might as well use copy-swap in this solution as well.

Reply:
That copy-and-swap was a bonus solution. The idiom is not really covered  by the text until Chapter 17, though.